### PR TITLE
Improve kill effect

### PR DESCRIPTION
**Pull request by EquipableMC**
### Description
This simply improves the kill effect by adding an option to prevent totems from popping

---
**Target Minecraft Versions:** any
**Requirements:** none
**Related Issues:** none

### CODE REVIEW

File: src/main/java/ch/njol/skript/effects/EffKill.java [hunk 1]
```diff
@@ -73,11 +75,14 @@ protected void execute(final Event e) {
 				entity = ((EnderDragonPart) entity).getParent();
 			}
 
+			if (IGNORE_TOTEM) {
+				HealthUtils.setHealth((Damageable) entity, 0);
```
Comment (Efnilite) [inline src/main/java/ch/njol/skript/effects/EffKill.java:79]: missing return? or else you set health to 0 and then do damage, or intended?

File: src/main/java/ch/njol/skript/effects/EffKill.java [hunk 2]
```diff
@@ -44,24 +44,26 @@
...
+		"kill all endermen, witches and bats",
+		"kill the player ignoring totem of undying",
+		"kill target entity ignoring the totem",
+		"kill target player ignoring resurrection"})
 @Since("1.0")
 public class EffKill extends Effect {
 
 	static {
-		Skript.registerEffect(EffKill.class, "kill %entities%");
+		Skript.registerEffect(EffKill.class, "kill %entities%", "kill %entities% ignoring [the] (resurrection|revival|[undying] totem|totem [of undying])");
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/effects/EffKill.java:55]: I'm not a huge fan of the syntax. I might prefer to just keep it simple:
```
ignoring [the|any] totem[s] [of undying]
```
I'd like to see what others think too though

Comment (EquipableMC) [reply src/main/java/ch/njol/skript/effects/EffKill.java:55]: What me and others have discussed on discord, is that the kill effect should be recoded to actually kill, since as of right now, it really just feels like a hardcoded damage effect. Another point is that the vanilla /kill command ignores the totem, whereas this, it does _not_ ignore the totem.

Comment (sovdeeth) [reply src/main/java/ch/njol/skript/effects/EffKill.java:55]: Pickle is talking about the pattern being overly complex, not about it's behavior.

Comment (Efnilite) [reply src/main/java/ch/njol/skript/effects/EffKill.java:55]: personally i think `ignoring [the|any] totem[s] of undying` should be it, as `ignoring totems` sounds too ambiguous imo

File: src/main/java/ch/njol/skript/effects/EffKill.java [hunk 3]
```diff
@@ -44,24 +44,26 @@
 		"Note: This effect does not set the entity's health to 0 (which causes issues), but damages the entity by 100 times its maximum health."})
 @Examples({"kill the player",
 		"kill all creepers in the player's world",
-		"kill all endermen, witches and bats"})
+		"kill all endermen, witches and bats",
+		"kill the player ignoring totem of undying",
+		"kill target entity ignoring the totem",
```
[the remark this answers is not in the document]
Comment (sovdeeth) [reply src/main/java/ch/njol/skript/effects/EffKill.java:31]: agreed

File: src/main/java/ch/njol/skript/effects/EffKill.java [hunk 4]
```diff
@@ -75,7 +80,7 @@ protected void execute(final Event e) {
 	}
 	
 	@Override
-	public String toString(final @Nullable Event e, final boolean debug) {
+	public String toString(@Nullable Event e, boolean debug) {
 		return "kill" + entities.toString(e, debug);
```
Comment (Efnilite) [inline src/main/java/ch/njol/skript/effects/EffKill.java:84]:
```suggestion
	public String toString(@Nullable Event event, boolean debug) {
		return "kill " + entities.toString(event, debug);
```

File: src/main/java/ch/njol/skript/effects/EffKill.java [hunk 5]
```diff
@@ -26,40 +26,45 @@
 		"Note: This effect does not set the entity's health to 0 (which causes issues), but damages the entity by 100 times its maximum health."})
 @Examples({"kill the player",
 		"kill all creepers in the player's world",
-		"kill all endermen, witches and bats"})
-@Since("1.0")
+		"kill all endermen, witches and bats",
+		"kill the player ignoring totem of undying",
+		"kill target entity ignoring the totem",
+		"kill target player ignoring resurrection"})
```
Comment (Efnilite) [inline src/main/java/ch/njol/skript/effects/EffKill.java:32]:
```suggestion
		"kill target player ignoring resurrection"
})
```

File: src/main/java/ch/njol/skript/effects/EffKill.java (hunk 1 above)
Comment (sovdeeth) [reply src/main/java/ch/njol/skript/effects/EffKill.java:61]: why was this resolved?

File: src/main/java/ch/njol/skript/effects/EffKill.java [hunk 6]
```diff
@@ -26,40 +26,45 @@
 		"Note: This effect does not set the entity's health to 0 (which causes issues), but damages the entity by 100 times its maximum health."})
 @Examples({"kill the player",
 		"kill all creepers in the player's world",
-		"kill all endermen, witches and bats"})
-@Since("1.0")
+		"kill all endermen, witches and bats",
+		"kill the player ignoring totem of undying",
+		"kill target entity ignoring the totem",
```
Comment (Efnilite) [inline src/main/java/ch/njol/skript/effects/EffKill.java:31]:
```suggestion
		"kill target entity ignoring totem of undying",
```

File: src/main/java/ch/njol/skript/effects/EffKill.java [hunk 7]
```diff
@@ -26,40 +26,45 @@
...
+		"kill target player ignoring resurrection"})
+@Since("1.0, INSERT VERSION (ignoring totem of undying)")
 public class EffKill extends Effect {
 
 	static {
-		Skript.registerEffect(EffKill.class, "kill %entities%");
+		Skript.registerEffect(EffKill.class, "kill %entities%", "kill %entities% ignoring [the|any] totem[s] of undying");
```
Comment (Efnilite) [inline src/main/java/ch/njol/skript/effects/EffKill.java:37]:
```suggestion
		Skript.registerEffect(EffKill.class, "kill %entities%", "kill %entities% ignoring [a|any] totem[s] of undying");
```

File: src/main/java/ch/njol/skript/effects/EffKill.java (hunk 1 above)
Comment (EquipableMC) [reply src/main/java/ch/njol/skript/effects/EffKill.java:61]: I lowkey don't know why, I must of forgotten to push changes the first time... oops

File: src/main/java/ch/njol/skript/effects/EffKill.java [hunk 8]
```diff
@@ -26,40 +26,45 @@
...
-	public boolean init(final Expression<?>[] exprs, final int matchedPattern, final Kleenean isDelayed, final ParseResult parser) {
+	public boolean init(Expression<?>[] exprs, int matchedPattern, Kleenean isDelayed, ParseResult parser) {
 		entities = (Expression<Entity>) exprs[0];
+		ignoreTotem = matchedPattern == 1;
 		return true;
 	}
 
 	@Override
-	protected void execute(final Event e) {
+	protected void execute(Event e) {
```
Comment (Efnilite) [inline src/main/java/ch/njol/skript/effects/EffKill.java:53]:
```suggestion
	protected void execute(Event event) {
```

**Review by Efnilite — CHANGES_REQUESTED** [review]

File: src/main/java/ch/njol/skript/effects/EffKill.java [hunk 9]
```diff
@@ -26,40 +26,46 @@
...
 	}
 
 	@Override
-	protected void execute(final Event e) {
-		for (Entity entity : entities.getArray(e)) {
+	protected void execute(Event event) {
+		for (Entity entity : entities.getArray(event)) {
 
 			if (entity instanceof EnderDragonPart) {
 				entity = ((EnderDragonPart) entity).getParent();
```
Comment (Efnilite) [inline src/main/java/ch/njol/skript/effects/EffKill.java:57]:
```suggestion
			if (entity instanceof EnderDragonPart part) {
				entity = part.getParent();
```

File: src/main/java/ch/njol/skript/effects/EffKill.java [hunk 10]
```diff
@@ -26,40 +26,46 @@
...
+			if (ignoreTotem) {
+				HealthUtils.setHealth((Damageable) entity, 0);
+				return;
+			}
+
 			if (entity instanceof Damageable) {
-				final boolean creative = entity instanceof Player && ((Player) entity).getGameMode() == GameMode.CREATIVE;
+				boolean creative = entity instanceof Player player && player.getGameMode() == GameMode.CREATIVE;
 				if (creative) // Set player to survival before applying damage
 					((Player) entity).setGameMode(GameMode.SURVIVAL);
```
Comment (Efnilite) [inline src/main/java/ch/njol/skript/effects/EffKill.java:68]:
```suggestion
					player.setGameMode(GameMode.SURVIVAL);
```

File: src/main/java/ch/njol/skript/effects/EffKill.java [hunk 11]
```diff
@@ -26,40 +26,46 @@
...
 			if (entity instanceof EnderDragonPart) {
 				entity = ((EnderDragonPart) entity).getParent();
 			}
 
+			if (ignoreTotem) {
+				HealthUtils.setHealth((Damageable) entity, 0);
+				return;
+			}
+
 			if (entity instanceof Damageable) {
```
Comment (Efnilite) [inline src/main/java/ch/njol/skript/effects/EffKill.java:65]:
```suggestion
			if (entity instanceof Damageable damageable) {
```

File: src/main/java/ch/njol/skript/effects/EffKill.java [hunk 12]
```diff
@@ -26,40 +26,46 @@
...
-				final boolean creative = entity instanceof Player && ((Player) entity).getGameMode() == GameMode.CREATIVE;
+				boolean creative = entity instanceof Player player && player.getGameMode() == GameMode.CREATIVE;
 				if (creative) // Set player to survival before applying damage
 					((Player) entity).setGameMode(GameMode.SURVIVAL);
-
 				HealthUtils.damage((Damageable) entity, HealthUtils.getMaxHealth((Damageable) entity) * 100); // just to make sure that it really dies >:)
```
Comment (Efnilite) [inline src/main/java/ch/njol/skript/effects/EffKill.java:69]:
```suggestion
				HealthUtils.damage(damageable, HealthUtils.getMaxHealth(damageable * 100); // just to make sure that it really dies >:)
```

File: src/main/java/ch/njol/skript/effects/EffKill.java (hunk 4 above)
Comment (Efnilite) [reply src/main/java/ch/njol/skript/effects/EffKill.java]: missing space

File: src/main/java/ch/njol/skript/effects/EffKill.java [hunk 13]
```diff
@@ -26,40 +26,46 @@
...
 
 				if (creative) // Set creative player back to creative
```
Comment (Efnilite) [inline src/main/java/ch/njol/skript/effects/EffKill.java:71]: line below this one can also have pattern matched var

**Review by Efnilite — CHANGES_REQUESTED** [review]

File: src/main/java/ch/njol/skript/effects/EffKill.java (hunk 10 above)
Comment (EquipableMC) [reply src/main/java/ch/njol/skript/effects/EffKill.java:68]: Cannot be changed, it does not accept `player`, as seen below
image

**Comment by EquipableMC** [comment]
I removed `ignoring the totem of undying` and removed setting the player to survival from creative, then back as that was unneeded, which is now making this a breaking change.

**Review by Efnilite — APPROVED** [review]
this will be a breaking change as the kill effect will now always kill, no matter if the player has a totem. in my opinion, this is expected behaviour, and therefore not a big deal

File: src/main/java/ch/njol/skript/effects/EffKill.java [hunk 14]
```diff
@@ -26,44 +26,35 @@
 		"Note: This effect does not set the entity's health to 0 (which causes issues), but damages the entity by 100 times its maximum health."})
 @Examples({"kill the player",
 		"kill all creepers in the player's world",
-		"kill all endermen, witches and bats"})
+		"kill all endermen, witches and bats",
+})
```
Comment (Moderocky) [inline src/main/java/ch/njol/skript/effects/EffKill.java:30]: Why did you add a `,` and spaces at the end? Please keep it consistent for both brackets.

**Review by Moderocky — CHANGES_REQUESTED** [review]

File: src/main/java/ch/njol/skript/effects/EffKill.java (hunk 14 above)
Comment (EquipableMC) [reply src/main/java/ch/njol/skript/effects/EffKill.java:30]: I thought I removed it, but I didn't realize I didn't, oopsies

File: src/main/java/ch/njol/skript/effects/EffKill.java [hunk 15]
```diff
@@ -26,44 +26,35 @@
...
 @Since("1.0")
 public class EffKill extends Effect {
 
 	static {
 		Skript.registerEffect(EffKill.class, "kill %entities%");
 	}
 	
-	// Absolutely make sure it dies
-	public static final int DAMAGE_AMOUNT = Integer.MAX_VALUE;
```
Comment (Sparky200) [inline src/main/java/ch/njol/skript/effects/EffKill.java:38]: Just a general rule in API, you should add this back with a deprecated and scheduled for removal annotation. It's highly unlikely this is used by any consumers of the API, but nevertheless good practice in case it does, so the consumer has more opportunity to change their use case

Comment (EquipableMC) [reply src/main/java/ch/njol/skript/effects/EffKill.java:38]: That was legacy code, and I thought I should remove it, but it really didn't matter, since this is like the least important place to apply it.

**Review by Sparky200 — APPROVED** [review]

**Review by sovdeeth — APPROVED** [review]

**Review by APickledWalrus — APPROVED** [review]

File: src/main/java/ch/njol/skript/effects/EffKill.java [hunk 16]
```diff
@@ -44,24 +44,26 @@
...
 
 	static {
-		Skript.registerEffect(EffKill.class, "kill %entities%");
+		Skript.registerEffect(EffKill.class, "kill %entities%", "kill %entities% ignoring [the] (resurrection|revival|[undying] totem|totem [of undying])");
 	}
 	
-	// Absolutely make sure it dies
-	public static final int DAMAGE_AMOUNT = Integer.MAX_VALUE;
-	
 	@SuppressWarnings("null")
```
[the remark this answers is not in the document]
Comment (Moderocky) [reply src/main/java/ch/njol/skript/effects/EffKill.java:36]: You need to actually do these changes rather than just clicking Resolve Conversation.

File: src/main/java/ch/njol/skript/effects/EffKill.java [hunk 17]
```diff
@@ -44,24 +44,26 @@
...
 	private Expression<Entity> entities;
+	private boolean IGNORE_TOTEM;
 	
 	@SuppressWarnings({"unchecked", "null"})
```
[the remark this answers is not in the document]
Comment (Moderocky) [reply src/main/java/ch/njol/skript/effects/EffKill.java:39]: See above.

**Review by Moderocky — CHANGES_REQUESTED** [review]

**Review by Moderocky — APPROVED** [review]
